File: src/core/spatial/flowField.ts
/**
 * THE FLOW FIELD - one search from the player that every enemy reads, instead of every enemy
 * working out for itself how to get round a wall.
 *
 * ---------------------------------------------------------------------------------------------
 * WHY THIS EXISTS: THE HORDE IS "ALL SOURCES, ONE DESTINATION"
 * ---------------------------------------------------------------------------------------------
 * Pathfinding problems are classified by how many starts and how many ends they have. A* is the
 * answer to ONE source and ONE destination. This game is the other shape entirely - three hundred
 * bodies, all heading for the same place - and for that the standard answer is to run a single
 * search FROM THE GOAL and let every agent read the result. It is called a flow field in RTS
 * circles (Supreme Commander 2, Planetary Annihilation) and a Dijkstra map in roguelike ones
 * (Brogue), and it is the same idea both times.
 *
 * What it replaces was a wall FOLLOWER - each body feeling its way round an obstacle from local
 * probes. That is the Bug family of algorithms from robotics, and it has a documented ceiling
 * rather than a fixable bug: local sensing cannot escape a local minimum, so there is always some
 * pocket a body settles in. Three rounds of fixes each solved one shape and failed on the next
 * (see the header of tests/enemyPathing.test.ts). A field cannot have that failure, because the
 * search that built it has already seen the whole neighbourhood.
 *
 * ---------------------------------------------------------------------------------------------
 * IT IS A WINDOW, AND THE WINDOW IS BIG ENOUGH BECAUSE THE HORDE IS ON A LEASH
 * ---------------------------------------------------------------------------------------------
 * Mossy Mayhem is unbounded, so a field over "the map" is not a thing that exists. It does not
 * need to be: `RELOCATE_RADIUS` (constants.ts) teleports any ordinary straggler that gets further
 * than 1000 u from the player back onto the spawn ring, so THE WHOLE HORDE LIVES IN A 1000 u DISC
 * whether it is on screen or not. FLOW_CELLS x FLOW_CELLS at 64 u is +-1536 u, which is that disc
 * with half as much again in margin.
 *
 * So "off screen" is not the same as "outside the field". An enemy behind a wall two screens away
 * is still inside the window, still gets a direction, and still arrives - which is the whole point
 * of a horde that is meant to apply pressure from every side.
 *
 * TWO THINGS ARE OUTSIDE IT, both deliberately. A BOSS is never relocated (outrunning a set-piece
 * is allowed to work), and anything else that has somehow got beyond the window. Both fall back to
 * the local follower in `seek`, which is exactly what a body with no better information should do.
 *
 * ---------------------------------------------------------------------------------------------
 * EVERYBODY DOES NOT TAKE THE SAME ROUTE
 * ---------------------------------------------------------------------------------------------
 * A field with one direction per cell steers every body in that cell identically, so a pack
 * rounding a wall files through the same gap one behind another. It does not have to: in a
 * distance field EVERY neighbour that is strictly closer is a valid route, because distance falls
 * by at least one each step and the walk therefore has to terminate. The search already computed
 * all of them and was keeping only the best.
 *
 * So `options` keeps the whole set and each body takes whichever member best matches its own fixed
 * lean on the bearing to the player (`flowDirFor`). Measured on the real lattice, 87.9% of
 * reachable cells offer two or more - the choice was always there, it was simply thrown away.
 *
 * ---------------------------------------------------------------------------------------------
 * COST, MEASURED
 * ---------------------------------------------------------------------------------------------
 * In a browser, on the real lattice, a full 48x48 rebuild is 96 us of cost-field sampling plus
 * 17 us of BFS. The whole simulation at 300 enemies is 496 us per tick against a 16 700 us frame,
 * so a rebuild is a fifth of one tick - and it does not happen every tick. See `updateFlowField`:
 * standing still it happens ONCE, and walking flat out about 254 times a minute, which is 8 us per
 * tick amortised. Squeezing it further was measured and rejected: at that share of a 700 us tick
 * there is nothing to win.
 *
 * IT IS CHEAPER THAN WHAT IT REPLACES. The per-enemy probing the follower did was measured by
 * removing it: 695 -> 641 us per tick at 300 enemies, so those probes cost ~54 us EVERY tick
 * against ~8 us amortised for the field.
 *
 * ---------------------------------------------------------------------------------------------
 * DERIVED STATE, SO IT IS NOT IN THE HASH
 * ---------------------------------------------------------------------------------------------
 * The field is a pure function of the terrain and the player's cell, exactly as the spatial hash
 * is a pure function of the enemy pool - and like the hash it is left out of `hashWorld`. What
 * must be deterministic is the ORDER things happen in, and it is: the BFS visits neighbours in a
 * fixed order, and the decision to rebuild reads only simulation state.
 */

import { sceneryOverlap, sceneryVersion } from '../content/scenery.js';
import type { World } from '../types.js';

/**
 * Edge of one field cell, world units. THE SAME 64 AS THE MOSSY WALL LATTICE, on purpose: on that
 * map a field cell is exactly a wall cell, so a wall is either in or out with no aliasing and no
 * doorway half-blocked by rounding. On the Scrapyard the piles do not sit on any grid, so there
 * the number is simply a resolution - and 64 u against a 45-90 u pile is fine.
 */
export const FLOW_CELL = 64;

/**
 * Window edge, in cells. 48 * 64 = 3072 u, so +-1536 u about the player.
 *
 * SIZED AGAINST `RELOCATE_RADIUS` (1000 u), not against the screen. See the header: the leash is
 * what bounds the horde, and a field that only covered the view would leave the bodies waiting
 * behind a wall off-screen with no way round - which is the pressure never arriving.
 */
export const FLOW_CELLS = 48;

/**
 * Nominal body the field is built for. Runt and grunt radii are 13-21; the field marks a cell
 * blocked if a body this size could not stand at its centre.
 *
 * ONE SIZE FOR EVERY BODY, deliberately. A per-radius field would be one search per distinct size
 * and the whole saving here is that there is exactly one search. A bruiser is a little wider than
 * this and will clip a corner the field routed it through - which the push-out then slides it off,
 * costing a moment rather than a path.
 */
const FLOW_BODY_RADIUS = 18;

/**
 * The eight directions a cell can point, as unit vectors. Diagonals are 1/sqrt(2) written out -
 * a literal, so nothing here depends on a runtime sqrt agreeing between engines.
 */
const D = 0.7071067811865476;
const DIR_X = Object.freeze([1, D, 0, -D, -1, -D, 0, D]);
const DIR_Y = Object.freeze([0, D, 1, D, 0, -D, -1, -D]);
/** Cell offsets matching DIR_X/DIR_Y. Index 1,3,5,7 are the diagonals. */
const OFF_X = Object.freeze([1, 1, 0, -1, -1, -1, 0, 1]);
const OFF_Y = Object.freeze([0, 1, 1, 1, 0, -1, -1, -1]);

export interface FlowField {
  /** Window origin, in field-cell coordinates. */
  originCx: number;
  originCy: number;
  /** 1 where a body could not stand. */
  readonly blocked: Uint8Array;
  /** Steps to the player, or -1 for unreachable and for cells the flood never got to. */
  readonly dist: Int32Array;
  /** Index into DIR_X/DIR_Y of the BEST descent, or -1 where there is nothing to point at. */
  readonly dir: Int8Array;
  /**
   * EVERY valid descent from this cell, as a bitmask over DIR_X/DIR_Y.
   *
   * This is what route variation is made of, and it costs one byte per cell because the search had
   * already worked it out and was throwing it away. A neighbour qualifies if its distance is
   * strictly LOWER, which is exactly the condition that makes following it safe: distance falls by
   * at least one every step, so any walk down any of these bits reaches the player in at most
   * `dist` moves. Measured on the real lattice, 87.9% of reachable cells have two or more bits set
   * and 86.2% have three - so the choice was always there.
   */
  readonly options: Uint8Array;
  /** BFS scratch, allocated once. */
  readonly queue: Int32Array;
  /** Rebuild bookkeeping. `builtTick` of -1 means "never built". */
  builtTick: number;
  builtCx: number;
  builtCy: number;
  /** `sceneryVersion` as of the last build. A change here means the terrain moved under us. */
  builtVersion: number;
  /** How many rebuilds this run. Diagnostics and the harness; nothing branches on it. */
  rebuilds: number;
}

export function createFlowField(): FlowField {
  const n = FLOW_CELLS * FLOW_CELLS;
  return {
    originCx: 0,
    originCy: 0,
    blocked: new Uint8Array(n),
    dist: new Int32Array(n),
    dir: new Int8Array(n),
    options: new Uint8Array(n),
    queue: new Int32Array(n),
    builtTick: -1,
    builtCx: 0,
    builtCy: 0,
    builtVersion: -1,
    rebuilds: 0,
  };
}

/** Field-cell index for a world coordinate, on either axis. */
export function flowCellOf(v: number): number {
  return Math.floor(v / FLOW_CELL);
}

/**
 * Rebuilds the field if what it was built from has changed. Called once a tick, between player
 * movement (S3) and the horde's steering (S4), so the goal is where the player is THIS tick.
 *
 * ---------------------------------------------------------------------------------------------
 * TWO TRIGGERS, AND BOTH ARE EXACT
 * ---------------------------------------------------------------------------------------------
 * The field is a pure function of two things - where the player is standing, and what the terrain
 * looks like - so it needs rebuilding exactly when one of those changes, and never otherwise.
 *
 * IT USED TO EXPIRE ON A TIMER INSTEAD, every half second, because there was no way to ask the
 * terrain whether it had moved. That was wrong in both directions at once: standing in an empty
 * field where nothing had changed it rebuilt 120 times a minute for nothing, and when a tree DID
 * come down it took up to half a second to notice - which is precisely the moment a route opens
 * and the horde ought to pour through it. `sceneryVersion` replaced the timer with the actual
 * question, so the wasted rebuilds are gone and the felled tree is seen on the very next tick.
 *
 * Both triggers read only simulation state, so two runs of the same seed rebuild on the same ticks.
 */
export function updateFlowField(world: World): void {
  const f = world.flow;
  const cx = flowCellOf(world.player.x);
  const cy = flowCellOf(world.player.y);
  const version = sceneryVersion(world.scenery);
  if (f.builtTick >= 0 && cx === f.builtCx && cy === f.builtCy && version === f.builtVersion) {
    return;
  }

  const N = FLOW_CELLS;
  const half = N >> 1;
  f.originCx = cx - half;
  f.originCy = cy - half;
  f.builtCx = cx;
  f.builtCy = cy;
  f.builtVersion = version;
  f.builtTick = world.tick;
  f.rebuilds++;

  const blocked = f.blocked;
  const dist = f.dist;
  const dir = f.dir;
  const options = f.options;
  const queue = f.queue;
  const scenery = world.scenery;

  // ---- 1. THE COST FIELD. The only place the terrain is touched, and where nearly all the time
  // goes - the flood over it afterwards is a fifth of the cost.
  for (let ry = 0; ry < N; ry++) {
    const wy = (f.originCy + ry + 0.5) * FLOW_CELL;
    const row = ry * N;
    for (let rx = 0; rx < N; rx++) {
      const wx = (f.originCx + rx + 0.5) * FLOW_CELL;
      blocked[row + rx] = sceneryOverlap(scenery, wx, wy, FLOW_BODY_RADIUS) >= 0 ? 1 : 0;
    }
  }

  // ---- 2. THE INTEGRATION FIELD: breadth-first out from the player's own cell.
  //
  // FOUR-WAY, not eight. A body may move diagonally - the direction pass below emits diagonals -
  // but the DISTANCES must not, because an eight-way flood makes a diagonal step as cheap as an
  // orthogonal one and the field then prefers staircases that hug corners. Four-way distances with
  // an eight-way descent is the standard pairing and it is what HowToRTS and Red Blob both use.
  dist.fill(-1);
  let head = 0;
  let tail = 0;
  const goal = half * N + half;
  // The player standing inside terrain would otherwise seed nothing and leave the whole field
  // unreachable, which reads as the horde giving up. Seed it regardless; the flood still spreads.
  dist[goal] = 0;
  queue[tail++] = goal;
  while (head < tail) {
    const i = queue[head++];
    const iy = (i / N) | 0;
    const ix = i - iy * N;
    const d = dist[i] + 1;
    if (ix > 0) {
      const j = i - 1;
      if (dist[j] < 0 && blocked[j] === 0) {
        dist[j] = d;
        queue[tail++] = j;
      }
    }
    if (ix < N - 1) {
      const j = i + 1;
      if (dist[j] < 0 && blocked[j] === 0) {
        dist[j] = d;
        queue[tail++] = j;
      }
    }
    if (iy > 0) {
      const j = i - N;
      if (dist[j] < 0 && blocked[j] === 0) {
        dist[j] = d;
        queue[tail++] = j;
      }
    }
    if (iy < N - 1) {
      const j = i + N;
      if (dist[j] < 0 && blocked[j] === 0) {
        dist[j] = d;
        queue[tail++] = j;
      }
    }
  }

  // ---- 3. THE FLOW FIELD: each reachable cell points at whichever of its eight neighbours is
  // nearest the player.
  //
  // Done ONCE PER REBUILD rather than per enemy per tick, which is the right way round: a rebuild
  // happens about three times a second and there are three hundred bodies reading it sixty times.
  //
  // A DIAGONAL IS REFUSED UNLESS BOTH ITS ORTHOGONALS ARE OPEN. Without that test a body cuts the
  // corner between two walls that meet at a point - the field says the diagonal is nearer, and it
  // is, but the body cannot fit through the join and grinds on it instead.
  for (let i = 0; i < N * N; i++) {
    if (dist[i] < 0) {
      dir[i] = -1;
      options[i] = 0;
      continue;
    }
    const iy = (i / N) | 0;
    const ix = i - iy * N;
    const here = dist[i];
    let best = -1;
    let bestD = here;
    let mask = 0;
    for (let k = 0; k < 8; k++) {
      const nx = ix + OFF_X[k];
      const ny = iy + OFF_Y[k];
      if (nx < 0 || nx >= N || ny < 0 || ny >= N) continue;
      const j = ny * N + nx;
      const nd = dist[j];
      if (nd < 0 || nd >= here) continue;
      if ((k & 1) === 1) {
        // Diagonal: both shoulders must be open.
        if (blocked[iy * N + nx] !== 0 || blocked[ny * N + ix] !== 0) continue;
      }
      // EVERY strictly-lower neighbour is a valid route, not just the lowest. The mask keeps all
      // of them; `dir` keeps the best, which is what a body with no preference gets.
      mask |= 1 << k;
      if (nd < bestD) {
        bestD = nd;
        best = k;
      }
    }
    dir[i] = best;
    options[i] = mask;
  }
}

/** Where to walk, written by `flowSteer`. Module scratch: `seek` allocates nothing. */
export let FLOW_X = 0;
export let FLOW_Y = 0;

/**
 * Where the body at (x, y) should walk, written into FLOW_X/FLOW_Y. False when the field has
 * nothing to say - outside the window, inside terrain, or in a pocket the flood never reached -
 * and the caller falls back to its own steering.
 *
 * ---------------------------------------------------------------------------------------------
 * THE CALLER DECIDES WHEN TO ASK, AND THAT TEST MUST BE LOCAL
 * ---------------------------------------------------------------------------------------------
 * It is tempting to have the field decide for itself whether a body needs it - compare the flood
 * distance against the straight-line distance, and only detour when the route is longer. That was
 * tried and it is WRONG, for a reason worth writing down because it looks correct:
 *
 *   THE FLOOD IS FOUR-WAY, so `dist == manhattan` means SOME monotone staircase to the player
 *   exists. It does not mean the STRAIGHT LINE is clear. Standing at the western end of a wall you
 *   can reach the player in Manhattan steps by going south first and then east - while the direct
 *   line still runs through the wall.
 *
 * The two then disagree across a cell boundary: one cell says "walk straight" (east, into the
 * wall) and its neighbour says "detour" (west, round it), and a body straddling the boundary
 * alternates between them forever. Measured - a body sat at the corner of an eight-cell wall for
 * a whole run, flipping east/west every few ticks, and 0 of 12 got round.
 *
 * So the trigger stays where it can see the truth: a short probe in front of the body, in `seek`.
 * When something is actually there, the field says where to go instead. The probe and the field
 * agree about the local facts, so there is nothing for them to oscillate between.
 */
export function flowDirAt(f: FlowField, x: number, y: number): boolean {
  if (f.builtTick < 0) return false;
  const rx = flowCellOf(x) - f.originCx;
  const ry = flowCellOf(y) - f.originCy;
  if (rx < 0 || rx >= FLOW_CELLS || ry < 0 || ry >= FLOW_CELLS) return false;
  const k = f.dir[ry * FLOW_CELLS + rx];
  if (k < 0) return false;
  FLOW_X = DIR_X[k];
  FLOW_Y = DIR_Y[k];
  return true;
}

/**
 * HOW EACH BODY LIKES TO COME AT YOU: four fixed rotations of the bearing to the player, as
 * cos/sin pairs.
 *
 * A body picks whichever valid descent best matches ITS bearing rather than the shortest one, so a
 * pack that used to file through a gap one behind another now arrives spread across the whole band
 * of routes that work. Two of the four lean each way, so the horde splits rather than drifting.
 *
 * +-40 and +-14 degrees. Wider than 40 and a body spends so long going sideways that it reads as
 * confused rather than as flanking; tighter than 14 and it is the same route as everybody else.
 * Written as literals - the rotation is two multiplies and there is no trigonometry at runtime to
 * disagree between V8 in CI and JSC on the phone.
 */
const SWIRL_COS = Object.freeze([
  0.766044443118978, 0.9702957262759965, 0.9702957262759965, 0.766044443118978,
]);
const SWIRL_SIN = Object.freeze([
  -0.6427876096865393, -0.24192189559966773, 0.24192189559966773, 0.6427876096865393,
]);

/**
 * The direction THIS body should take out of its cell, given where it wants to go and who it is.
 *
 * ---------------------------------------------------------------------------------------------
 * WHY EVERY CHOICE HERE IS SAFE
 * ---------------------------------------------------------------------------------------------
 * It picks from `options`, and every bit in that mask is a neighbour whose distance to the player
 * is STRICTLY LOWER. So whichever one a body takes, its distance falls by at least one cell per
 * step and it arrives in at most `dist` moves. There is no route in here that can loop, dead-end
 * or wander - the variation is over paths that all work, which is the whole point.
 *
 * THE COST OF THE SCENIC ROUTE IS BOUNDED, and worth stating because "any descending neighbour"
 * sounds looser than it is. The best step drops the distance by 2 (a diagonal) and the worst by 1,
 * so a body that takes the scenic option every single time walks at most twice the cells of one
 * that never does. In practice it is far less, because the preferred bearing keeps pulling back
 * toward the player.
 *
 * `id` is the body's `spawnId`: stable for its whole life, so its choice does not flicker tick to
 * tick, and consecutive spawns get different leanings so a wave fans out instead of forming a line.
 */
export function flowDirFor(
  f: FlowField,
  x: number,
  y: number,
  ux: number,
  uy: number,
  id: number,
): boolean {
  if (f.builtTick < 0) return false;
  const rx = flowCellOf(x) - f.originCx;
  const ry = flowCellOf(y) - f.originCy;
  if (rx < 0 || rx >= FLOW_CELLS || ry < 0 || ry >= FLOW_CELLS) return false;
  const mask = f.options[ry * FLOW_CELLS + rx];
  if (mask === 0) return false;

  // The bearing this body would like to be travelling on: straight at the player, turned by its
  // own fixed lean.
  const s = id & 3;
  const c = SWIRL_COS[s];
  const sn = SWIRL_SIN[s];
  const wx = ux * c - uy * sn;
  const wy = ux * sn + uy * c;

  let best = -1;
  let bestDot = -Infinity;
  for (let k = 0; k < 8; k++) {
    if ((mask & (1 << k)) === 0) continue;
    const dot = DIR_X[k] * wx + DIR_Y[k] * wy;
    // Strictly greater, so ties fall to the lowest k and the result cannot depend on iteration
    // order changing under us.
    if (dot > bestDot) {
      bestDot = dot;
      best = k;
    }
  }
  if (best < 0) return false;
  FLOW_X = DIR_X[best];
  FLOW_Y = DIR_Y[best];
  return true;
}

File: cs/src/Scrapyard.Core/FlowField.cs
namespace Scrapyard.Core;

/// <summary>
/// The field the horde steers by: a breadth-first flood out from the player, rebuilt only when it
/// has gone stale. A port of <c>src/core/spatial/flowField.ts</c>.
/// </summary>
/// <remarks>
/// <para>
/// NOT HASHED, deliberately, and for a reason worth stating precisely: it is a pure function of the
/// player's cell and the scenery version, both of which ARE hashed. A divergence here becomes a
/// divergence in enemy positions within a tick or two. See the rule at the top of
/// <see cref="Hash.HashWorld"/>.
/// </para>
/// <para>
/// That does NOT make it safe to get wrong. It makes it a thing whose failure surfaces one tick
/// late and a long way from the cause, which is why it gets a fixture of its own rather than being
/// left to the corpus.
/// </para>
/// </remarks>
public sealed class FlowField
{
    public const int Cell = 64;

    /// <summary>48 cells at 64 u is +-1536 u about the player, which covers the camera's reach.</summary>
    public const int Cells = 48;

    /// <summary>
    /// The radius a body is assumed to have when deciding whether a cell is passable. One number
    /// for the whole horde: a per-body field would be 300 floods a tick.
    /// </summary>
    private const double BodyRadius = 18;

    private const double D = 0.7071067811865476;

    private static readonly double[] DirX = { 1, D, 0, -D, -1, -D, 0, D };
    private static readonly double[] DirY = { 0, D, 1, D, 0, -D, -1, -D };
    private static readonly int[] OffX = { 1, 1, 0, -1, -1, -1, 0, 1 };
    private static readonly int[] OffY = { 0, 1, 1, 1, 0, -1, -1, -1 };

    /// <summary>
    /// A fixed per-body LEAN, four of them, indexed by the low two bits of the spawn id.
    /// </summary>
    /// <remarks>
    /// Precomputed cos/sin pairs rather than an angle, so nothing in the per-body path ever calls a
    /// trig function - the same reason <c>rotateTowardsInto</c> takes a pair. Four values means a
    /// crowd converging on the player arrives as a swirl rather than as a single column, and a
    /// body's lean never changes, so it does not oscillate between two routes.
    /// </remarks>
    private static readonly double[] SwirlCos =
    {
        0.766044443118978, 0.9702957262759965, 0.9702957262759965, 0.766044443118978,
    };

    private static readonly double[] SwirlSin =
    {
        -0.6427876096865393, -0.24192189559966773, 0.24192189559966773, 0.6427876096865393,
    };

    /// <summary>tan(22.5 degrees). The half-width of an octant, for snapping a bearing to one of eight.</summary>
    private const double Octant = 0.41421356237309503;

    public int OriginCx;
    public int OriginCy;
    public readonly byte[] Blocked;
    public readonly int[] Dist;
    public readonly sbyte[] Dir;

    /// <summary>Bitmask of every strictly-lower neighbour, not just the best one.</summary>
    public readonly byte[] Options;

    private readonly int[] _queue;

    public int BuiltTick = -1;
    public int BuiltCx;
    public int BuiltCy;
    public int BuiltVersion = -1;
    public int Rebuilds;

    public FlowField()
    {
        int n = Cells * Cells;
        Blocked = new byte[n];
        Dist = new int[n];
        Dir = new sbyte[n];
        Options = new byte[n];
        _queue = new int[n];
    }

    /// <summary>World coordinate to flow cell. Floor, not truncation - the arena straddles 0.</summary>
    public static int CellOf(double v) => (int)Math.Floor(v / Cell);

    /// <summary>
    /// Rebuilds the field if the player has moved to a new cell or the terrain has changed.
    /// </summary>
    /// <remarks>
    /// The staleness test is the whole reason this is affordable: about three rebuilds a second
    /// against three hundred bodies reading it sixty times.
    /// </remarks>
    public void Update(World w, IScenery scenery, double playerX, double playerY)
    {
        int cx = CellOf(playerX);
        int cy = CellOf(playerY);
        int version = scenery.Version;
        if (BuiltTick >= 0 && cx == BuiltCx && cy == BuiltCy && version == BuiltVersion) return;

        const int n = Cells;
        int half = n >> 1;
        OriginCx = cx - half;
        OriginCy = cy - half;
        BuiltCx = cx;
        BuiltCy = cy;
        BuiltVersion = version;
        BuiltTick = w.Tick;
        Rebuilds++;

        // ---- 1. THE COST FIELD. The only place terrain is touched, and where nearly all the time
        // goes - the flood over it afterwards is a fifth of the cost.
        for (int ry = 0; ry < n; ry++)
        {
            double wy = (OriginCy + ry + 0.5) * Cell;
            int row = ry * n;
            for (int rx = 0; rx < n; rx++)
            {
                double wx = (OriginCx + rx + 0.5) * Cell;
                Blocked[row + rx] = (byte)(scenery.Overlap(wx, wy, BodyRadius) >= 0 ? 1 : 0);
            }
        }

        // ---- 2. THE INTEGRATION FIELD: breadth-first out from the player's own cell.
        //
        // FOUR-WAY, not eight. A body MAY move diagonally - the direction pass below emits
        // diagonals - but the DISTANCES must not, because an eight-way flood makes a diagonal step
        // as cheap as an orthogonal one, and the field then prefers staircases that hug corners.
        // Four-way distances with an eight-way descent is the standard pairing.
        Array.Fill(Dist, -1);
        int head = 0;
        int tail = 0;
        int goal = half * n + half;

        // The player standing INSIDE terrain would otherwise seed nothing and leave the whole field
        // unreachable, which reads as the horde giving up. Seed it regardless; the flood spreads.
        Dist[goal] = 0;
        _queue[tail++] = goal;

        while (head < tail)
        {
            int i = _queue[head++];
            int iy = i / n;
            int ix = i - iy * n;
            int d = Dist[i] + 1;

            // Unrolled in the same order as the TypeScript - left, right, up, down. The ORDER
            // decides nothing about the distances (BFS is order-independent for those) but it does
            // decide the queue's contents, and `Rebuilds` and the queue are observable.
            if (ix > 0) Visit(i - 1, d, ref tail);
            if (ix < n - 1) Visit(i + 1, d, ref tail);
            if (iy > 0) Visit(i - n, d, ref tail);
            if (iy < n - 1) Visit(i + n, d, ref tail);
        }

        // ---- 3. THE FLOW FIELD: each reachable cell points at whichever of its eight neighbours
        // is nearest the player. Done ONCE PER REBUILD rather than per enemy per tick.
        for (int i = 0; i < n * n; i++)
        {
            if (Dist[i] < 0)
            {
                Dir[i] = -1;
                Options[i] = 0;
                continue;
            }

            int iy = i / n;
            int ix = i - iy * n;
            int here = Dist[i];
            int best = -1;
            int bestD = here;
            int mask = 0;

            for (int k = 0; k < 8; k++)
            {
                int nx = ix + OffX[k];
                int ny = iy + OffY[k];
                if (nx < 0 || nx >= n || ny < 0 || ny >= n) continue;
                int j = ny * n + nx;
                int nd = Dist[j];
                if (nd < 0 || nd >= here) continue;

                if ((k & 1) == 1)
                {
                    // DIAGONAL: both shoulders must be open. Without this a body cuts the corner
                    // between two walls that meet at a point - the field says the diagonal is
                    // nearer, and it is, but the body cannot fit through the join and grinds on it.
                    if (Blocked[iy * n + nx] != 0 || Blocked[ny * n + ix] != 0) continue;
                }

                // EVERY strictly-lower neighbour is a valid route, not just the lowest. The mask
                // keeps all of them; `Dir` keeps the best, which is what a body with no preference
                // gets.
                mask |= 1 << k;
                if (nd < bestD)
                {
                    bestD = nd;
                    best = k;
                }
            }

            Dir[i] = (sbyte)best;
            Options[i] = (byte)mask;
        }
    }

    private void Visit(int j, int d, ref int tail)
    {
        if (Dist[j] < 0 && Blocked[j] == 0)
        {
            Dist[j] = d;
            _queue[tail++] = j;
        }
    }

    /// <summary>
    /// WOULD WALKING STRAIGHT AT THE PLAYER ACTUALLY GET ANY CLOSER? False in open ground, true
    /// when the straight line runs into something.
    /// </summary>
    /// <remarks>
    /// The cheap test that keeps the field off the hot path: a body only consults the flow when the
    /// direct bearing is not a route, which in open ground is never.
    /// </remarks>
    public bool Detours(double x, double y, double ux, double uy)
    {
        if (BuiltTick < 0) return false;
        int rx = CellOf(x) - OriginCx;
        int ry = CellOf(y) - OriginCy;
        if (rx < 0 || ry < 0 || rx >= Cells || ry >= Cells) return false;
        int here = Dist[ry * Cells + rx];

        // -1 is unreachable (the field has no route to offer from here) and 0 is the player's own
        // cell. Neither is a detour.
        if (here <= 0) return false;

        // The neighbour the straight bearing steps into - the bearing snapped to one of the eight.
        double ax = ux < 0 ? -ux : ux;
        double ay = uy < 0 ? -uy : uy;
        int ox = 0;
        int oy = 0;
        if (ay <= ax * Octant) ox = ux < 0 ? -1 : 1;
        else if (ax <= ay * Octant) oy = uy < 0 ? -1 : 1;
        else
        {
            ox = ux < 0 ? -1 : 1;
            oy = uy < 0 ? -1 : 1;
        }

        int nx = rx + ox;
        int ny = ry + oy;
        if (nx < 0 || ny < 0 || nx >= Cells || ny >= Cells) return false;
        int there = Dist[ny * Cells + nx];

        // Walled off, or no closer than standing still: either way the straight line is not a route.
        return there < 0 || there >= here;
    }

    /// <summary>
    /// The step a specific body should take: the flow, leaned by that body's own swirl.
    /// </summary>
    /// <remarks>
    /// <para>
    /// AIMED AT THE CENTRE OF THE CELL BEING STEPPED INTO, not along a bare axis, and this is the
    /// half of the field that is easy to leave out.
    /// </para>
    /// <para>
    /// <c>Blocked</c> is sampled at CELL CENTRES: a cell is open when a body could stand in the
    /// middle of it. A body is almost never in the middle of anything. One hugging the top edge of
    /// its cell, handed a bare "west", walks west along that edge - and its radius-18 circle clips
    /// the corner of the building diagonally north-west, a cell the four-way flood never had to
    /// consider because the step it approved was orthogonal. The push-out then removes exactly the
    /// westward component, the velocity comes out as precisely zero, and the next tick sets up the
    /// identical frame. Measured in the original: a body parked against a courtyard wall for 20 s
    /// with a clear gateway two cells away, the field pointing at it the whole time, velocity
    /// (0, 0) every tick.
    /// </para>
    /// <para>
    /// Aiming at the target cell's centre folds re-centring into the steering for free. A body
    /// already centred gets the bare axis back, so nothing changes in open ground.
    /// </para>
    /// </remarks>
    public bool DirFor(double x, double y, double ux, double uy, int id, out Vec2 dir)
    {
        dir = default;
        if (BuiltTick < 0) return false;
        int rx = CellOf(x) - OriginCx;
        int ry = CellOf(y) - OriginCy;
        if (rx < 0 || rx >= Cells || ry < 0 || ry >= Cells) return false;
        int mask = Options[ry * Cells + rx];
        if (mask == 0) return false;

        // The bearing this body would like to travel on: straight at the player, turned by its own
        // fixed lean.
        int sIdx = id & 3;
        double c = SwirlCos[sIdx];
        double sn = SwirlSin[sIdx];
        double wx = ux * c - uy * sn;
        double wy = ux * sn + uy * c;

        int best = -1;
        double bestDot = double.NegativeInfinity;
        for (int k = 0; k < 8; k++)
        {
            if ((mask & (1 << k)) == 0) continue;
            double dot = DirX[k] * wx + DirY[k] * wy;
            // STRICTLY greater, so ties fall to the lowest k and the result cannot depend on
            // iteration order changing under us.
            if (dot > bestDot)
            {
                bestDot = dot;
                best = k;
            }
        }
        if (best < 0) return false;

        double tx = (OriginCx + rx + OffX[best] + 0.5) * Cell;
        double ty = (OriginCy + ry + OffY[best] + 0.5) * Cell;
        double vx = tx - x;
        double vy = ty - y;
        double l2 = vx * vx + vy * vy;

        // Dead centre of the target cell: no vector exists, and the axis is the same answer anyway.
        if (l2 == 0)
        {
            dir.X = DirX[best];
            dir.Y = DirY[best];
            return true;
        }

        double inv = 1 / Math.Sqrt(l2);
        dir.X = vx * inv;
        dir.Y = vy * inv;
        return true;
    }

    /// <summary>
    /// The unit step a body at (x, y) should take, or false when the cell has no route.
    /// </summary>
    /// <remarks>
    /// Returns into <paramref name="dir"/> rather than allocating, matching the original's
    /// module-level out-parameters - except that those are module-scoped in the TypeScript and this
    /// is a plain out, which is strictly safer: two worlds stepped in one process cannot collide.
    /// </remarks>
    public bool DirAt(double x, double y, out Vec2 dir)
    {
        dir = default;
        if (BuiltTick < 0) return false;
        int rx = CellOf(x) - OriginCx;
        int ry = CellOf(y) - OriginCy;
        if (rx < 0 || rx >= Cells || ry < 0 || ry >= Cells) return false;
        int k = Dir[ry * Cells + rx];
        if (k < 0) return false;
        dir.X = DirX[k];
        dir.Y = DirY[k];
        return true;
    }
}
